File: alchemist/alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/implementations/nodes/GenericNode.java
/*
 * Copyright (C) 2010-2014, Danilo Pianini and contributors
 * listed in the project's pom.xml file.
 * 
 * This file is part of Alchemist, and is distributed under the terms of
 * the GNU General Public License, with a linking exception, as described
 * in the file LICENSE in the Alchemist distribution's top directory.
 */
/**
 * 
 */
package it.unibo.alchemist.model.implementations.nodes;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Spliterator;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.Semaphore;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;

import org.danilopianini.concurrency.ThreadLocalIdGenerator;

import com.google.common.collect.MapMaker;
import com.googlecode.concurrentlinkedhashmap.ConcurrentLinkedHashMap;

import it.unibo.alchemist.model.interfaces.Environment;
import it.unibo.alchemist.model.interfaces.Molecule;
import it.unibo.alchemist.model.interfaces.Node;
import it.unibo.alchemist.model.interfaces.Reaction;


/**
 * This class realizes an abstract node. You may extend it to realize your own
 * nodes.
 * 
 * @param <T>
 */
public abstract class GenericNode<T> implements Node<T> {

    private static final long serialVersionUID = 2496775909028222278L;
    private static final ConcurrentMap<Environment<?>, AtomicInteger> IDGENERATOR = new MapMaker()
            .weakKeys().makeMap();
    private static final Semaphore MUTEX = new Semaphore(1);
    private static final ThreadLocalIdGenerator SINGLETON = new ThreadLocalIdGenerator();
    private static final AtomicInteger THREAD_UNSAFE = new AtomicInteger();
    private final int id;
    private final List<Reaction<T>> reactions = new ArrayList<>();
    private final ConcurrentMap<Molecule, T> molecules = new ConcurrentLinkedHashMap.Builder<Molecule, T>()
            .maximumWeightedCapacity(Long.MAX_VALUE)
            .concurrencyLevel(2)
            .build();

    private static int idFromEnv(final Environment<?> env) {
        MUTEX.acquireUninterruptibly();
        AtomicInteger idgen = IDGENERATOR.get(Objects.requireNonNull(env));
        if (idgen == null) {
            idgen = new AtomicInteger();
            IDGENERATOR.put(env, idgen);
        }
        MUTEX.release();
        return idgen.getAndIncrement();
    }

    /**
     * Basically, builds the node and just caches the hash code.
     * 
     * @param threadLocal
     *            true if the id should be local to the current thread. In order
     *            to keep the node ids along multiple simulations, pass true and
     *            use different threads to instance the nodes.
     * @deprecated this constructor can not generate ids correctly in case
     *             batches of simulation are executed by the same thread, and as
     *             such it has been deprecated and is scheduled for removal at
     *             the next major release.
     */
    @Deprecated
    protected GenericNode(final boolean threadLocal) {
        this(threadLocal ? SINGLETON.genId() : THREAD_UNSAFE.getAndIncrement());
    }

    /**
     * @param env
     *            the environment, used to generate sequential ids for each
     *            environment, always starting from 0.
     */
    public GenericNode(final Environment<?> env) {
        this(idFromEnv(env));
    }

    private GenericNode(final int id) {
        this.id = id;
    }

    @Override
    public void addReaction(final Reaction<T> r) {
        reactions.add(r);
    }

    @Override
    public GenericNode<T> cloneNode() {
        throw new UnsupportedOperationException();
    }

    @Override
    public int compareTo(final Node<T> o) {
        if (o instanceof GenericNode<?>) {
            if (id > ((GenericNode<?>) o).id) {
                return 1;
            }
            if (id < ((GenericNode<?>) o).id) {
                return -1;
            }
        }
        return 0;
    }

    @Override
    public boolean contains(final Molecule m) {
        return molecules.containsKey(m);
    }

    /**
     * @return an empty concentration
     */
    protected abstract T createT();

    @Override
    public boolean equals(final Object o) {
        if (o instanceof GenericNode<?>) {
            return ((GenericNode<?>) o).id == id;
        }
        return false;
    }

    @Override
    public void forEach(final Consumer<? super Reaction<T>> action) {
        reactions.forEach(action);
    }

    @Override
    public int getChemicalSpecies() {
        return molecules.size();
    }

    @Override
    public T getConcentration(final Molecule mol) {
        final T res = molecules.get(mol);
        if (res == null) {
            return createT();
        }
        return res;
    }

    @Override
    public Map<Molecule, T> getContents() {
        return Collections.unmodifiableMap(molecules);
    }

    @Override
    public int getId() {
        return id;
    }

    @Override
    public List<Reaction<T>> getReactions() {
        return reactions;
    }

    @Override
    public int hashCode() {
        return id;
    }

    @Override
    public Iterator<Reaction<T>> iterator() {
        return reactions.iterator();
    }

    @Override
    public void removeConcentration(final Molecule mol) {
        molecules.remove(mol);
    }

    @Override
    public void removeReaction(final Reaction<T> r) {
        reactions.remove(r);
    }

    @Override
    public void setConcentration(final Molecule mol, final T c) {
        molecules.put(mol, c);
    }

    @Override
    public Spliterator<Reaction<T>> spliterator() {
        return reactions.spliterator();
    }

    @Override
    public String toString() {
        return molecules.toString();
    }

}

File: alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/implementations/nodes/GenericNode.kt
/*
 * Copyright (C) 2010-2022, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.implementations.nodes

import com.google.common.collect.MapMaker
import it.unibo.alchemist.model.interfaces.Environment
import it.unibo.alchemist.model.interfaces.Incarnation
import it.unibo.alchemist.model.interfaces.Molecule
import it.unibo.alchemist.model.interfaces.Node
import it.unibo.alchemist.model.interfaces.NodeProperty
import it.unibo.alchemist.model.interfaces.Reaction
import it.unibo.alchemist.model.interfaces.Time
import java.util.Collections
import java.util.Spliterator
import java.util.concurrent.Semaphore
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer
import javax.annotation.Nonnull

/**
 * This class realizes an abstract node. You may extend it to realize your own
 * nodes.
 *
 * @param <T> concentration type
</T> */
open class GenericNode<T> @JvmOverloads constructor(
    /**
     * simulation incarnation.
     */
    val incarnation: Incarnation<T, *>,
    /**
     * The environment in which the node is places.
     */
    val environment: Environment<T, *>,
    override val id: Int = idFromEnv(environment),
    override val reactions: MutableList<Reaction<T>> = ArrayList(),
    /**
     * The node's molecules.
     */
    val molecules: MutableMap<Molecule, T> = LinkedHashMap(),
    override val properties: MutableList<NodeProperty<T>> = ArrayList(),
) : Node<T> {

    constructor(
        environment: Environment<T, *>,
    ) : this(environment.incarnation, environment)

    final override fun addReaction(reactionToAdd: Reaction<T>) {
        reactions.add(reactionToAdd)
    }

    override fun cloneNode(currentTime: Time): Node<T> = GenericNode(environment).also {
        this.properties.forEach { property -> it.addProperty(property.cloneOnNewNode(it)) }
        this.contents.forEach(it::setConcentration)
        this.reactions.forEach { reaction -> it.addReaction(reaction.cloneOnNewNode(it, currentTime)) }
    }

    final override fun compareTo(@Nonnull other: Node<T>): Int = id.compareTo(other.id)

    override fun contains(molecule: Molecule): Boolean = molecules.containsKey(molecule)

    /**
     * @return an empty concentration
     */
    protected open fun createT(): T = incarnation.createConcentration()

    final override fun equals(other: Any?): Boolean = other is Node<*> && other.id == id

    /**
     * Performs an [action] for every reaction.
     */
    final override fun forEach(action: Consumer<in Reaction<T>>) = reactions.forEach(action)

    override fun getConcentration(molecule: Molecule): T = molecules[molecule] ?: createT()

    override val contents: Map<Molecule, T> = Collections.unmodifiableMap(molecules)

    override val moleculeCount: Int get() = molecules.size

    final override fun hashCode(): Int = id // TODO: better hashing

    final override fun iterator(): Iterator<Reaction<T>> = reactions.iterator()

    final override fun removeConcentration(moleculeToRemove: Molecule) {
        if (molecules.remove(moleculeToRemove) == null) {
            throw NoSuchElementException("$moleculeToRemove was not present in node $id")
        }
    }

    final override fun removeReaction(reactionToRemove: Reaction<T>) {
        reactions.remove(reactionToRemove)
    }

    override fun setConcentration(molecule: Molecule, concentration: T) {
        molecules[molecule] = concentration
    }

    final override fun addProperty(nodeProperty: NodeProperty<T>) {
        if (properties.find { it::class == nodeProperty::class } == null) {
            properties.add(nodeProperty)
        } else {
            throw IllegalArgumentException(
                "This node (${this.id}) already contains a property of type ${nodeProperty::class}," +
                    "this may lead to an inconsistent state"
            )
        }
    }

    /**
     * Returns the [reactions] [Spliterator].
     */
    final override fun spliterator(): Spliterator<Reaction<T>> = reactions.spliterator()

    override fun toString(): String = molecules.toString()

    companion object {
        private const val serialVersionUID = 2496775909028222278L

        private val IDGENERATOR = MapMaker().weakKeys().makeMap<Environment<*, *>, AtomicInteger>()

        private val MUTEX = Semaphore(1)

        private fun idFromEnv(environment: Environment<*, *>): Int {
            MUTEX.acquireUninterruptibly()
            var idgen = IDGENERATOR[environment]
            if (idgen == null) {
                idgen = AtomicInteger()
                IDGENERATOR[environment] = idgen
            }
            MUTEX.release()
            return idgen.getAndIncrement()
        }
    }
}
